#include "common_dbc.h"

namespace {

const Signal sigs_37[] = {
    {
      .name = "TOYOTA_STEER_ANGLE",
      .b1 = 4,
      .b2 = 12,
      .bo = 48,
      .is_signed = true,
      .factor = 1.5,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "TOYOTA_STEER_RATE",
      .b1 = 36,
      .b2 = 12,
      .bo = 16,
      .is_signed = true,
      .factor = 1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "TOYOTA_STEER_FRACTION",
      .b1 = 32,
      .b2 = 4,
      .bo = 28,
      .is_signed = true,
      .factor = 0.1,
      .offset = 0,
      .is_little_endian = false,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_512[] = {
    {
      .name = "CHECKSUM",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_CHECKSUM,
    },
    {
      .name = "COUNTER",
      .b1 = 40,
      .b2 = 4,
      .bo = 20,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_COUNTER,
    },
    {
      .name = "GAS_COMMAND",
      .b1 = 8,
      .b2 = 16,
      .bo = 40,
      .is_signed = false,
      .factor = 0.275676,
      .offset = -33.3567,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "GAS_COMMAND2",
      .b1 = 24,
      .b2 = 16,
      .bo = 24,
      .is_signed = false,
      .factor = 0.13629,
      .offset = -32.71,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "ENABLE",
      .b1 = 47,
      .b2 = 1,
      .bo = 16,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_513[] = {
    {
      .name = "CHECKSUM",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_CHECKSUM,
    },
    {
      .name = "COUNTER",
      .b1 = 40,
      .b2 = 4,
      .bo = 20,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_COUNTER,
    },
    {
      .name = "PED_GAS",
      .b1 = 8,
      .b2 = 16,
      .bo = 40,
      .is_signed = false,
      .factor = 0.275676,
      .offset = -33.3567,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "PED_GAS2",
      .b1 = 24,
      .b2 = 16,
      .bo = 24,
      .is_signed = false,
      .factor = 0.13629,
      .offset = -32.71,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "STATE",
      .b1 = 44,
      .b2 = 4,
      .bo = 16,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_526[] = {
    {
      .name = "CHECKSUM",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_CHECKSUM,
    },
    {
      .name = "COUNTER",
      .b1 = 8,
      .b2 = 4,
      .bo = 52,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_COUNTER,
    },
    {
      .name = "BRAKE_MODE",
      .b1 = 12,
      .b2 = 2,
      .bo = 50,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "BRAKE_RELATIVE_COMMAND",
      .b1 = 16,
      .b2 = 16,
      .bo = 32,
      .is_signed = false,
      .factor = 0.0078125,
      .offset = -252.0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "BRAKE_POSITION_COMMAND",
      .b1 = 32,
      .b2 = 12,
      .bo = 20,
      .is_signed = false,
      .factor = 0.015625,
      .offset = -5.0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_527[] = {
    {
      .name = "CHECKSUM",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_CHECKSUM,
    },
    {
      .name = "COUNTER",
      .b1 = 8,
      .b2 = 4,
      .bo = 52,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_COUNTER,
    },
    {
      .name = "STATUS",
      .b1 = 12,
      .b2 = 4,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "BRAKE_OK",
      .b1 = 16,
      .b2 = 1,
      .bo = 47,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "DRIVER_BRAKE_APPLIED",
      .b1 = 17,
      .b2 = 1,
      .bo = 46,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "BRAKE_APPLIED",
      .b1 = 18,
      .b2 = 1,
      .bo = 45,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "BRAKE_PEDAL_POSITION",
      .b1 = 20,
      .b2 = 12,
      .bo = 32,
      .is_signed = false,
      .factor = 0.015625,
      .offset = -5.0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "EXT_STATUS1",
      .b1 = 32,
      .b2 = 4,
      .bo = 28,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "EXT_STATUS2",
      .b1 = 36,
      .b2 = 4,
      .bo = 24,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_608[] = {
    {
      .name = "CHECKSUM",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_CHECKSUM,
    },
    {
      .name = "COUNTER",
      .b1 = 8,
      .b2 = 4,
      .bo = 52,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_COUNTER,
    },
    {
      .name = "STATUS",
      .b1 = 12,
      .b2 = 4,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "STEERING_OK",
      .b1 = 16,
      .b2 = 1,
      .bo = 47,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "STEERING_TORQUE_DRIVER",
      .b1 = 18,
      .b2 = 11,
      .bo = 35,
      .is_signed = true,
      .factor = 0.8,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "STEERING_TORQUE_EPS",
      .b1 = 29,
      .b2 = 11,
      .bo = 24,
      .is_signed = true,
      .factor = 0.8,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_741[] = {
    {
      .name = "CHECKSUM",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_CHECKSUM,
    },
    {
      .name = "COUNTER",
      .b1 = 8,
      .b2 = 4,
      .bo = 52,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_COUNTER,
    },
    {
      .name = "STEER_MODE",
      .b1 = 12,
      .b2 = 2,
      .bo = 50,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "REQUESTED_STEER_ANGLE",
      .b1 = 14,
      .b2 = 15,
      .bo = 35,
      .is_signed = true,
      .factor = 0.1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "REQUESTED_STEER_TORQUE",
      .b1 = 29,
      .b2 = 11,
      .bo = 24,
      .is_signed = true,
      .factor = 0.5,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_869[] = {
    {
      .name = "CHECKSUM",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_CHECKSUM,
    },
    {
      .name = "SET_BTN",
      .b1 = 8,
      .b2 = 1,
      .bo = 55,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "CANCEL_BTN",
      .b1 = 9,
      .b2 = 1,
      .bo = 54,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SPEEDUP_BTN",
      .b1 = 10,
      .b2 = 1,
      .bo = 53,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SPEEDDN_BTN",
      .b1 = 11,
      .b2 = 1,
      .bo = 52,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "STATE",
      .b1 = 12,
      .b2 = 4,
      .bo = 48,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};
const Signal sigs_870[] = {
    {
      .name = "CHECKSUM",
      .b1 = 0,
      .b2 = 8,
      .bo = 56,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::OCELOT_CHECKSUM,
    },
    {
      .name = "ENABLED",
      .b1 = 8,
      .b2 = 1,
      .bo = 55,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "SET_SPEED",
      .b1 = 16,
      .b2 = 8,
      .bo = 40,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
    {
      .name = "CURRENT_SPEED",
      .b1 = 24,
      .b2 = 8,
      .bo = 32,
      .is_signed = false,
      .factor = 1,
      .offset = 0,
      .is_little_endian = true,
      .type = SignalType::DEFAULT,
    },
};

const Msg msgs[] = {
  {
    .name = "TOYOTA_STEERING_ANGLE_SENSOR1",
    .address = 0x25,
    .size = 8,
    .num_sigs = ARRAYSIZE(sigs_37),
    .sigs = sigs_37,
  },
  {
    .name = "GAS_COMMAND",
    .address = 0x200,
    .size = 6,
    .num_sigs = ARRAYSIZE(sigs_512),
    .sigs = sigs_512,
  },
  {
    .name = "GAS_SENSOR",
    .address = 0x201,
    .size = 6,
    .num_sigs = ARRAYSIZE(sigs_513),
    .sigs = sigs_513,
  },
  {
    .name = "BRAKE_COMMAND",
    .address = 0x20E,
    .size = 6,
    .num_sigs = ARRAYSIZE(sigs_526),
    .sigs = sigs_526,
  },
  {
    .name = "BRAKE_STATUS",
    .address = 0x20F,
    .size = 5,
    .num_sigs = ARRAYSIZE(sigs_527),
    .sigs = sigs_527,
  },
  {
    .name = "STEERING_STATUS",
    .address = 0x260,
    .size = 5,
    .num_sigs = ARRAYSIZE(sigs_608),
    .sigs = sigs_608,
  },
  {
    .name = "STEERING_COMMAND",
    .address = 0x2E5,
    .size = 5,
    .num_sigs = ARRAYSIZE(sigs_741),
    .sigs = sigs_741,
  },
  {
    .name = "HIM_CTRLS",
    .address = 0x365,
    .size = 3,
    .num_sigs = ARRAYSIZE(sigs_869),
    .sigs = sigs_869,
  },
  {
    .name = "CURRENT_STATE",
    .address = 0x366,
    .size = 4,
    .num_sigs = ARRAYSIZE(sigs_870),
    .sigs = sigs_870,
  },
};

const Val vals[] = {
    {
      .name = "ENABLE",
      .address = 0x200,
      .def_val = "1 YES 0 NO",
      .sigs = sigs_512,
    },
    {
      .name = "STATE",
      .address = 0x201,
      .def_val = "5 FAULT_TIMEOUT 4 FAULT_STARTUP 3 FAULT_SCE 2 FAULT_SEND 1 FAULT_BAD_CHECKSUM 0 NO_FAULT",
      .sigs = sigs_513,
    },
    {
      .name = "BRAKE_MODE",
      .address = 0x20E,
      .def_val = "3 RESERVED 2 POSITION_CONTROL 1 RELATIVE_CONTROL 0 DISENGADGED",
      .sigs = sigs_526,
    },
    {
      .name = "BRAKE_OK",
      .address = 0x20F,
      .def_val = "1 YES 0 NO",
      .sigs = sigs_527,
    },
    {
      .name = "IBOOSTER_BRAKE_APPLIED",
      .address = 0x20F,
      .def_val = "1 YES 0 NO",
      .sigs = sigs_527,
    },
    {
      .name = "STATUS",
      .address = 0x20F,
      .def_val = "5 FAULT_TIMEOUT 4 FAULT_STARTUP 3 FAULT_SCE 2 FAULT_SEND 1 FAULT_BAD_CHECKSUM 0 NO_FAULT",
      .sigs = sigs_527,
    },
    {
      .name = "STATUS",
      .address = 0x260,
      .def_val = "5 FAULT_TIMEOUT 4 FAULT_STARTUP 3 FAULT_SCE 2 FAULT_SEND 1 FAULT_BAD_CHECKSUM 0 NO_FAULT",
      .sigs = sigs_608,
    },
    {
      .name = "STEERING_OK",
      .address = 0x260,
      .def_val = "1 YES 0 NO",
      .sigs = sigs_608,
    },
    {
      .name = "STEER_MODE",
      .address = 0x2E5,
      .def_val = "3 RESERVED 2 POSITION_CONTROL 1 TORQUE_CONTROL 0 DISENGADGED",
      .sigs = sigs_741,
    },
    {
      .name = "CANCEL_BTN",
      .address = 0x365,
      .def_val = "1 YES 0 NO",
      .sigs = sigs_869,
    },
    {
      .name = "SET_BTN",
      .address = 0x365,
      .def_val = "1 YES 0 NO",
      .sigs = sigs_869,
    },
    {
      .name = "SPEEDDN_BTN",
      .address = 0x365,
      .def_val = "1 YES 0 NO",
      .sigs = sigs_869,
    },
    {
      .name = "SPEEDUP_BTN",
      .address = 0x365,
      .def_val = "1 YES 0 NO",
      .sigs = sigs_869,
    },
    {
      .name = "STATE",
      .address = 0x365,
      .def_val = "5 FAULT_TIMEOUT 4 FAULT_STARTUP 3 FAULT_SCE 2 FAULT_SEND 1 FAULT_BAD_CHECKSUM 0 NO_FAULT",
      .sigs = sigs_869,
    },
    {
      .name = "ENABLED",
      .address = 0x366,
      .def_val = "1 YES 0 NO",
      .sigs = sigs_870,
    },
};

}

const DBC ocelot_can = {
  .name = "ocelot_can",
  .num_msgs = ARRAYSIZE(msgs),
  .msgs = msgs,
  .vals = vals,
  .num_vals = ARRAYSIZE(vals),
};

dbc_init(ocelot_can)